watertap/unit_models/pressure_changer.py

```diff
@@ -11,35 +11,301 @@
 import idaes.core.util.scaling as iscale
 
 import idaes.logger as idaeslog
 
 _log = idaeslog.getLogger(__name__)
 
 
+class VariableEfficiency(Enum):
+    none = auto()  # default is constant efficiency
+    flow = auto()  # flow-only correlation
+    flow_head = auto()  # flow and head correlation
+
+
 @declare_process_block_class("Pump")
 class PumpIsothermalData(PumpData):
     """
     Standard Isothermal Pump Unit Model Class
     """
 
+    CONFIG = PumpData.CONFIG()
+    CONFIG.declare(
+        "variable_efficiency",
+        ConfigValue(
+            default=VariableEfficiency.none,
+            domain=In(VariableEfficiency),
+            description="Variable pump efficiency flag",
+            doc="""Indicates the relationship used to define pump efficiency
+    **VariableEfficiency.none** - uses default pump efficiency at BEP
+    **VariableEfficiency.flow** - uses an efficiency correlation scaled to the BEP flow rate
+    **VariableEfficiency.flow_head** - uses an efficiency correlation scaled to the BEP flow rate and head
+    """,
+        ),
+    )
+
+    def build(self):
+        super().build()
+
+        # ---------------------------------------
+        # Isothermal pump set up
+        # ---------------------------------------
+
+        self.control_volume.del_component(self.control_volume.enthalpy_balances)
+
+        @self.control_volume.Constraint(
+            self.flowsheet().config.time, doc="Isothermal constraint"
+        )
+        def isothermal_balance(b, t):
+            return b.properties_in[t].temperature == b.properties_out[t].temperature
+
+        # ---------------------------------------
+        # Variable efficiency pump set-up
+        # ---------------------------------------
+
+        if self.config.variable_efficiency is not VariableEfficiency.none:
+            # create additional pyomo variables
+            self.bep_flow = Var(
+                initialize=1.0,
+                doc="Best efficiency point flowrate of the centrifugal pump",
+                units=pyunits.m**3 / pyunits.s,
+            )
+
+            self.bep_eta = Var(
+                initialize=self.efficiency_pump[0].value,
+                doc="Best efficiency of the centrifugal pump",
+                units=pyunits.dimensionless,
+            )
+
+            self.flow_ratio = Var(
+                self.flowsheet().time,
+                initialize=1.0,
+                doc="Ratio of pump flowrate to best efficiency point flowrate",
+                units=pyunits.dimensionless,
+            )
+
+            # unfix the default efficiency from the base class
+            self.efficiency_pump.unfix()
+
+            # add constraints
+            @self.Constraint(self.flowsheet().time, doc="Pump flow ratio")
+            def flow_ratio_constraint(b, t):
+                return b.flow_ratio[t] * b.bep_flow == (
+                    b.control_volume.properties_out[t].flow_vol.expr()
```

Comment:
Do we need the `.expr()`? I would think not.

Reply:
Removed.

Reply:
I second this - I do not see a reason why the `.expr()` should be required (and many good reasons why it should not).

For one, this assumes that `flow_vol` is an `Expression`, which maybe true in the current case, but is hardly general (and even then it should not be needed).